File: Sorting/MergeSort/mergesort.cpp
//
//  TYPE: Merge Sort[O(nlogn)] : https://www.acmicpc.net/problem/2751
//  GOAL: Sort Array in non-decreasing order
//  Paradigm: Divide_and_Conquer
#include <iostream>
#include <fstream>

#define INF 123456789

using namespace std;

void merge_sort(int* arr, int p, int r);
void merge(int* arr, int p, int q, int r, int s);

int main(int argc, char* argv[])
{
	cin.sync_with_stdio(false);
	ifstream in("input.txt", ifstream::in);
	if (!in.is_open()) { cout << "file open failed!\n"; return -1; }
	cin.rdbuf(in.rdbuf());

	int N; cin >> N;
	int* arr = new int[N];
	for (register int i = 0; i < N; ++i) cin >> arr[i];

	merge_sort(arr, 0, N - 1); // arr[0...N-1]
	
	for (register int i = 0; i < N; ++i) cout << arr[i] << "\n";

	delete[] arr;
	return 0;
}

void merge_sort(int* arr, int p, int r) // arr[p....r]
{
	// Base Condition
	if (p >= r) return; // 1-element
	// Divide
	int q = (p + r) / 2;
	merge_sort(arr, p, q);
	merge_sort(arr, q + 1, r);
	// Conquer
	merge(arr, p, q, q + 1, r);

	return;
}

void merge(int* arr, int p, int q, int r, int s) // arr[p...q] , arr[r...s]
{
	// Duplication Array for Merge
	int lhsIdx = 0, rhsIdx = 0;
	int* lhsArr = new int[q - p + 2];
	int* rhsArr = new int[s - r + 2];
	for (lhsIdx = 0; lhsIdx <= q - p; ++lhsIdx)
		lhsArr[lhsIdx] = arr[p + lhsIdx];
	lhsArr[lhsIdx] = INF;
	for (rhsIdx = 0; rhsIdx <= s - r; ++rhsIdx)
		rhsArr[rhsIdx] = arr[r + rhsIdx];
	rhsArr[rhsIdx] = INF;
	
	// Compare and Merge
	lhsIdx = rhsIdx = 0;
	for (register int k = p; k <= s; ++k) {
		if (lhsArr[lhsIdx] <= rhsArr[rhsIdx])
			arr[k] = lhsArr[lhsIdx++];
		else
			arr[k] = rhsArr[rhsIdx++];
	}
	delete[] lhsArr;
	delete[] rhsArr;

	return;
}

File: Sorting/MergeSort/merge_sort.py
#  https://www.acmicpc.net/problem/2750
#  TYPE: Merge Sort[O(nlogn)]
#  GOAL: Sort Array in non-decreasing order
#  Paradigm: Divide_and_Conque
import math, random, sys
sys.setrecursionlimit(10 ** 6)
sys.stdin.readline


def merge_sort(arr: list, p: int, r: int):
    # base condition
    if p >= r:
        return
    # divide
    q = (p + r) // 2
    merge_sort(arr, p, q)
    merge_sort(arr, q+1, r)
    # conquer
    merge(arr, p, q, q+1, r)
    return


def merge(arr: list, p: int, q: int, r: int, s: int):
    # lhs: left-hand side
    lhs_arr = arr[p:q+1]
    lhs_arr.append(math.inf)
    # rhs: right-hand side
    rhs_arr = arr[r:s+1]
    rhs_arr.append(math.inf)

    # compare and merge
    lhs_idx = rhs_idx = 0
    for k in range(p, s + 1):
        if lhs_arr[lhs_idx] <= rhs_arr[rhs_idx]:
            arr[k] = lhs_arr[lhs_idx]
            lhs_idx += 1
        else:
            arr[k] = rhs_arr[rhs_idx]
            rhs_idx += 1


n = int(input())
arr = []
for _ in range(n):
    num = int(input())
    arr.append(num)
merge_sort(arr, 0, n - 1)
for num in arr:
    print(num)
